--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/modules/removed/RemovedConfig.java
@@ -1,0 +_,22 @@
+package me.earthme.luminol.config.modules.removed;
+
+import me.earthme.luminol.config.*;
+
+public class RemovedConfig implements IConfigModule {
+    @TransformedConfig(name = "example", category = {"removed", "example"}, transform = false)
+    @DoNotLoad
+    @ConfigInfo(baseName = "removed", comments =
+            """
+                    RemovedConfig redirect to here, no any function.""")
+    public static boolean enabled = true;
+
+    @Override
+    public EnumConfigCategory getCategory() {
+        return EnumConfigCategory.REMOVED;
+    }
+
+    @Override
+    public String getBaseName() {
+        return "removed_config";
+    }
+}
