--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/TransformedConfig.java
@@ -1,0 +_,22 @@
+package me.earthme.luminol.config;
+
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Repeatable(TransformedConfig.List.class)
+public @interface TransformedConfig {
+    String name() default "";
+
+    String[] category() default "";
+
+    boolean transform() default true;
+
+    Class<? extends DefaultTransformLogic>[] transformLogic() default {DefaultTransformLogic.class};
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        TransformedConfig[] value();
+    }
+}
