From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Wed, 2 Jul 2025 22:39:21 +0800
Subject: [PATCH] Fix incorrect configuration mode checks

A simple fix of: https://github.com/PaperMC/Folia/issues/378

diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 31101f5af05772a0842b56ecc2dd8c79eebee933..4c36ecb1cd37dc564e12cc1f211d4976b44ef83b 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -98,6 +98,22 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         }
         // Folia start - region threading
         if (!(this instanceof ServerGamePacketListenerImpl gamePacketListener)) {
+            // Luminol start - Fix incorrect configuration mode checks
+            // we are not starting to turn into a game packet handler, so we can safely return
+            if (this instanceof ServerConfigurationPacketListenerImpl && !io.papermc.paper.threadedregions.RegionizedServer.isNotOwnedByGlobalRegion(this.connection)) {
+                // folia's weird entity scheduler process?
+                if (this.connection.savedPlayerForLoginEventLegacy != null) {
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueTickTaskQueue(
+                            this.connection.savedPlayerForLoginEventLegacy.level(),
+                            this.connection.savedPlayerForLoginEventLegacy.getBlockX() >> 4,
+                            this.connection.savedPlayerForLoginEventLegacy.getBlockZ() >> 4,
+                            () -> {
+                                this.connection.savedPlayerForLoginEventLegacy.getBukkitEntity().taskScheduler.retire();
+                            });
+                }
+                return;
+            }
+            // Luminol end
             throw new UnsupportedOperationException("Configuration mode is not supported in its current state");
         }
         gamePacketListener.player.getBukkitEntity().taskScheduler.retire();
@@ -356,14 +372,22 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
             return;
         }
         // Folia start - region threading
-        if (!(this instanceof ServerGamePacketListenerImpl gamePacketListener)) {
-            throw new UnsupportedOperationException("Configuration mode is not supported in its current state");
-        }
-        // Folia end - region threading
-        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(gamePacketListener.player)) { // Folia - region threading
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            // Luminol start - Fix incorrect configuration mode checks
+            // folia's weird entity scheduler process?
+            // we'll retire soonly
+            //throw new UnsupportedOperationException("Configuration mode is not supported in its current state");
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(gamePacketListener.player)) { // Folia - region threading
+                this.connection.disconnectSafely(disconnectionDetails); // Folia - region threading - it HAS to be delayed/async to avoid deadlock if we try to wait for another region
+                return;
+            }
+        }else {
+            // calling on configuration mode
             this.connection.disconnectSafely(disconnectionDetails); // Folia - region threading - it HAS to be delayed/async to avoid deadlock if we try to wait for another region
             return;
         }
+        // Folia end - region threading
+        // Luminol end
 
         Component reason;
         Component leaveMessage;
