From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 6 Apr 2025 10:42:47 +0800
Subject: [PATCH] Leaves Fix SculkCatalyst exp skip


diff --git a/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java b/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
index ab7584873e46020148bceecbd42a43055684e6a0..f33ab20313310a113d6c88e0fc1fe1666218061e 100644
--- a/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
+++ b/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
@@ -18,6 +18,7 @@ public class PlayerDeathEvent extends EntityDeathEvent {
     private boolean keepLevel = false;
     private boolean keepInventory = false;
     private boolean doExpDrop; // Paper - shouldDropExperience API
+    private boolean useApiExpDropStatus = false; // Leaves - exp fix
     // Paper start - adventure
     @org.jetbrains.annotations.ApiStatus.Internal
     public PlayerDeathEvent(final @NotNull Player player, final @NotNull DamageSource damageSource, final @NotNull List<ItemStack> drops, final int droppedExp, final @Nullable net.kyori.adventure.text.Component deathMessage) {
@@ -122,9 +123,16 @@ public class PlayerDeathEvent extends EntityDeathEvent {
      */
     public void setShouldDropExperience(boolean doExpDrop) {
         this.doExpDrop = doExpDrop;
+        this.useApiExpDropStatus = true; // Leaves - exp fix
     }
     // Paper end - shouldDropExperience API
 
+    // Leaves start - exp fix
+    public boolean forceUseEventDropStatus() {
+        return this.useApiExpDropStatus;
+    }
+    // Leaves end - exp fix
+
     @NotNull
     @Override
     public Player getEntity() {
